frame/nfts/src/features/atomic_swap.rs

```diff
@@ -0,0 +1,152 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::*;
+use frame_support::{
+	pallet_prelude::*,
+	traits::{Currency, ExistenceRequirement::KeepAlive},
+};
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	pub fn do_create_swap(
+		sender: T::AccountId,
+		collection_id: T::CollectionId,
+		item_id: T::ItemId,
+		desired_collection_id: T::CollectionId,
+		desired_item_id: T::ItemId,
+		maybe_price: Option<ItemPrice<T, I>>,
+		maybe_duration: Option<<T as SystemConfig>::BlockNumber>,
+	) -> DispatchResult {
+		let item = Item::<T, I>::get(&collection_id, &item_id).ok_or(Error::<T, I>::UnknownItem)?;
+		ensure!(item.owner == sender, Error::<T, I>::NoPermission);
+
+		Item::<T, I>::get(&desired_collection_id, &desired_item_id)
+			.ok_or(Error::<T, I>::UnknownItem)?;
+
+		let now = frame_system::Pallet::<T>::block_number();
+		let deadline = maybe_duration.map(|d| d.saturating_add(now));
+
+		PendingSwapOf::<T, I>::insert(
+			&collection_id,
+			&item_id,
+			PendingSwap {
+				desired_collection: desired_collection_id,
+				desired_item: desired_item_id,
+				price: maybe_price,
+				deadline,
+			},
+		);
+
+		Self::deposit_event(Event::SwapCreated {
+			collection: collection_id,
+			item: item_id,
+			desired_collection: desired_collection_id,
+			desired_item: desired_item_id,
+			price: maybe_price,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	pub fn do_cancel_swap(
+		sender: T::AccountId,
+		collection_id: T::CollectionId,
+		item_id: T::ItemId,
+	) -> DispatchResult {
+		let item = Item::<T, I>::get(&collection_id, &item_id).ok_or(Error::<T, I>::UnknownItem)?;
+		let swap = PendingSwapOf::<T, I>::get(&collection_id, &item_id)
+			.ok_or(Error::<T, I>::UnknownSwap)?;
+
+		let is_past_deadline = if let Some(deadline) = swap.deadline {
+			let now = frame_system::Pallet::<T>::block_number();
+			now > deadline
+		} else {
+			false
+		};
+
+		if !is_past_deadline {
+			ensure!(item.owner == sender, Error::<T, I>::NoPermission);
+		}
```

Comment:
Yes, right. Anyone could cancel an outdated swap and clear the storage

Reply:
But there is no incentive for it.

Reply:
> But there is no incentive for it.

Curious how this is usually tackled. I see two options
1. Make it worth user's while
2. Have an on_idle job clearing outdated swaps
3. Both?

Reply:
@ruseinov I saw deposit for the storage recored in Asset pallet. The amount stored within the same record.

Reply:
yeah, we either need to add deposits for everything data related, so this would incentivise users to clear the data.
But at the same time, the deposit would go back to the owner and not the method's caller. So, I don't see an incentive for the external user here to call this method.
I think this approach is mostly suitable for the `on_idle` job that would clear outdated data.